peps/pep-0765.rst

```diff
@@ -0,0 +1,169 @@
+than that which was proposed by :pep:`601`.
+
+Motivation
+==========
+
+The semantics of ``return``, ``break`` and ``continue`` in a
+finally block are surprising for many developers.
+The documentation [1]_ mentions that
+
+1. If the ``finally`` clause executes a ``break``, ``continue``
+or ``return`` statement, exceptions are not re-raised.
+2. If a ``finally`` clause includes a ``return`` statement, the
+returned value will be the one from the ``finally`` clause’s
+``return`` statement, not the value from the ``try`` clause’s
+``return`` statement.
+
+Both of these behaviours cause confusion, but the first is
+particularly dangerous because a swallowed exception is more
+likely to slip through testing, than an incorrect return value.
+
+In 2019, :pep:`601` proposed to change Python to emit a
+``SyntaxWarning`` for a few releases and then turn it into a
+``SyntaxError``. It was rejected in favour of viewing this
+as a programming style issue, to be handled by linters and :pep:`8`.
+Indeed, :pep:`8` now recommends not to use control flow statements
+in a ``finally`` block, and linters such as pylint [2]_, ruff [3]_
+and flake8-bugbear [4]_ flag them as a problem.
+
+Rationale
+=========
+
+A recent analysis of real world code [5]_ shows that:
+
+- These features are rare (2 per million LOC in the top 8000 PyPI
+  packages, 4 per million LOC in a random selection of packages).
+  This could be thanks to the linters that flag this pattern.
+- Most of the usages are incorrect, and introduce unintended
+  exception-swallowing bugs.
+- Code owners are typically receptive to fixing the bugs, and
+  find that easy to do.
+
+This new data indicates that it would benefit Python's users if
+Python itself moved them away from this harmful feature.
+
+One of the arguments brought up [6]_ in the discussion about :pep:`601`
+was that language features should be orthogonal, and combine without
+context-based restrictions. However, in the meantime :pep:`654` has
+been implemented, and it forbids ``return``, ``break`` and ``continue``
+in an ``except*`` clause because the semantics of that would violate
+the property that ``except*`` clauses operate *in parallel*, so the
+code of one clause should not suppress the invocation of another.
+In that case we accepted that a combination of features can be
+harmful enough that it makes sense to disallow it.
+
+
+Specification
+=============
+
+The change is to specify as part of the language spec that
+Python's compiler may emit a ``SyntaxWarning`` or ``SyntaxError``
+when a ``return``, ``break`` or ``continue`` would transfer
+control flow from within a ``finally`` block to a location outside
+of it.
+
+This includes the following examples:
+
+.. code-block::
+
+    def f():
+        try:
+            ...
+        finally:
+            return 42
+
+    for x in o:
+        try:
+            ...
+        finally:
+            break # (or continue)
+
+But excludes these:
+
+.. code-block::
```

Comment:
And this would give a green sidebar:
```suggestion
.. code-block::
   :class: good
```

(There's also `:class: maybe` for yellow: https://peps.python.org/pep-0008/#maximum-line-length)